pkg/api/response/response.go

```diff
@@ -80,6 +81,11 @@ func (r *NormalResponse) ErrMessage() string {
 
 func (r *NormalResponse) WriteTo(ctx *contextmodel.ReqContext) {
 	if r.err != nil {
+		grafanaErr := errutil.Error{}
+		if errors.As(r.err, &grafanaErr) && grafanaErr.Reason.Status().IsDownstream() {
+			requestmeta.WithDownstreamStatusSource(ctx.Req.Context())
```

Comment:
Current approach in regards to errors while dropping the idea of plugin client middleware. This is the main thing enabling errors to signal their originating from downstream. 

@xnyo @bergquist WDYT?

Reply:
`CallResource` API handler is special in that case it doesn't return anything, it uses the `contextmodel.ReqContext` to write the response since CallResource responses from plugins are streamed, ref
https://github.com/grafana/grafana/blob/f4c127a1d8aa0acbfb0c724eb1ed6a4f5a43b751/pkg/api/plugin_resource.go#L24

`QueryData` on the other hand returns a `response.Response`, see below, which is the reason I did the opposite here to reuse the response package logic by creating a `response.Response` and calling `WriteTo(ctx)` (the other way around :) )
https://github.com/grafana/grafana/blob/124f445db0e0738d14211461ffb76bf0afd8a4ec/pkg/api/metrics.go#L53

Reply:
There's still quite a lot of API handlers not returning a `response.Response`, but that's due to legacy reasons.